group/assignment-5/src/alphabet/UnknownAlphabet.java

```diff
@@ -0,0 +1,140 @@
+        // throws an exception with an appropraite message if the given dictionary is invalid
+        AlphabetDAG alphabetDAG = AlphabetDAG.buildGraphFromDictionary(dictionary);
+
+        return topologicalSort(alphabetDAG);
+    }
+
+    /**
+     * This method takes a list of strings representing words in a dictionary
+     * and returns all alphabets that can be formed from these words.
+     *
+     * @param dictionary This is a list containing all the words that form the alphabet
+     * @return List<List       <       Character>> This method returns a list of all possible alphabets
+     * @throws IllegalArgumentException
+     */
+    public static List<List<Character>> findAlphabets(List<String> dictionary)
+            throws IllegalArgumentException {
+        AlphabetDAG alphabetDAG = AlphabetDAG.buildGraphFromDictionary(dictionary);
+
+        List<List<Character>> alphabets = new ArrayList<>();
+        Set<AlphabetDAG.CharNode> inDegreeZero = new LinkedHashSet<>();
+        Map<AlphabetDAG.CharNode, Integer> nodesToInDegree = new HashMap<>();
+        updateCollectionsByInDegree(alphabetDAG, nodesToInDegree, inDegreeZero);
+
+        allTopologicalSorts(alphabets, inDegreeZero, nodesToInDegree, new ArrayList<>());
+        return alphabets;
+    }
+
+    private static void allTopologicalSorts(List<List<Character>> alphabets, Set<AlphabetDAG.CharNode> inDegreeZero, Map<AlphabetDAG.CharNode, Integer> nodesToInDegree, List<Character> curr) {
+        AlphabetDAG.CharNode currCharNode;
+        int size = inDegreeZero.size();
+        for (int i = 0; i < size; i++) {
+            currCharNode = inDegreeZero.iterator().next();
+            inDegreeZero.remove(currCharNode);
+            curr.add(currCharNode.getValue());
+            // reduce inDegree of forward neighbours
+            for (AlphabetDAG.CharNode neighbours : currCharNode.getForwardNeighbours()) {
+                int inDegree = nodesToInDegree.get(neighbours);
+                if (inDegree > 1) {
+                    nodesToInDegree.put(neighbours, inDegree - 1);
+                } else {
+                    nodesToInDegree.remove(neighbours);
+                    inDegreeZero.add(neighbours);
+                }
+            }
+            // go deeper
+            allTopologicalSorts(alphabets, inDegreeZero, nodesToInDegree, curr);
+            inDegreeZero.add(currCharNode);
+            curr.remove(curr.size() - 1);
+            // restore inDegree of neighbours
+            for (AlphabetDAG.CharNode neighbours : currCharNode.getForwardNeighbours()) {
+                int inDegree = nodesToInDegree.getOrDefault(neighbours, 0);
+                if (inDegree == 0) {
+                    inDegreeZero.remove(neighbours);
+                }
+                nodesToInDegree.put(neighbours, inDegree + 1);
+            }
+        }
+        if (inDegreeZero.isEmpty()) {
+            alphabets.add(curr.stream().collect(Collectors.toList()));
```

Comment:
```suggestion
            alphabets.add(curr);
```

curr is already a list

Reply:
since it's in a recursion we need to add a copy of curr because after that curr will change